backup-stores/s3/src/main/java/io/camunda/zeebe/backup/s3/S3BackupStore.java

```diff
@@ -66,14 +73,17 @@ public S3BackupStore(final S3BackupConfig config, final S3AsyncClient client) {
 
   @Override
   public CompletableFuture<Void> save(final Backup backup) {
-    return setStatus(backup.id(), Status.inProgress())
+    return requireBackupStatus(backup.id(), EnumSet.of(BackupStatusCode.DOES_NOT_EXIST))
+        .thenComposeAsync(id -> setStatus(id, Status.inProgress()))
         .thenComposeAsync(
             status -> {
               final var metadata = saveMetadata(backup);
               final var snapshot = saveSnapshotFiles(backup);
               final var segments = saveSegmentFiles(backup);
               return CompletableFuture.allOf(metadata, snapshot, segments);
             })
+        .thenComposeAsync(
+            ignored -> requireBackupStatus(backup.id(), EnumSet.of(BackupStatusCode.IN_PROGRESS)))
```

Comment:
:thought_balloon:  I'm wondering if this is enough because this is not an atomic read-check-update operation. Would something that is marked as failed, would get updated as completed? Probably then it is not that bad because it is marked as completed because it is actually completed.

Reply:
:question: Would it get marked as completed, after it is deleted? It could execute in following order due to concurrent operations: check status in progress -> mark as failed -> delete -> set status to complete

Reply:
Absolutely, that's a major issue that I tried to mitigate a little bit by checking the status at key moments but as you said, it doesn't solve it. I've already spent some time to look into possible solutions but as far as I know there's nothing waterproof and portable.